### ISSUE TITLE
Problem reaching normal endpoint under load

### ISSUE THREAD

**Issue by Jacobh2** [miguelgrinberg/Flask-SocketIO#805]
I've created a simple flask-socketio server and then using a socketio client to push data to this server. (Client is running Python 3.7.0 and using [socketIO-client](https://pypi.org/project/socketIO-client/))

I have this deployed in a kubernetes cluster with both readiness and liveness probes to the `index` method.

Client is very simple and just sending an increasing number and a uuid in a loop:
```py
i = 0
while True:
    try:
        socket_client.emit('data', {'data': i, 'id': str(uuid4())})
    except Exception:
        self.logger.warning("Crash while sending data!", exc_info=True)
        break
    i += 1
    sleep(0.001)
```

My problem is that I'm apparently sending data so fast to the server that the kublet fails to get a response from the index route, thus labeling my service dead and will restart it, disrupting the socket connection.

How can I make sure that all requests made to the server is being handled? Is eventlet not the way to go? I read a lot of other issues for example about gunicorn, but since that only supports 1 worker I'm guessing I'd still have this issue?

Server code:
```py
from flask import Flask, jsonify, request
from flask_socketio import SocketIO

app = Flask(__name__)
app.config['SECRET_KEY'] = 'secret!'
socketio = SocketIO(app, async_mode='eventlet', engineio_logger=True)

@app.route('/')
def index():
    # Used for readiness and liveness prob in kubernetes
    print("Yes, I'm alive!")
    return jsonify({'ok': True, 'msg': 'Test server'})

@socketio.on('data')
def on_data(json_data):
    sid = request.sid
    uid = json_data['id']
    data = json_data['data']

if __name__ == '__main__':
    socketio.run(app, host='0.0.0.0', port=80, debug=True, use_reloader=False)
```
And pip freeze:
```
Click==7.0
dnspython==1.15.0
eventlet==0.24.1
Flask==1.0.2
Flask-SocketIO==3.0.2
greenlet==0.4.15
itsdangerous==0.24
Jinja2==2.10
MarkupSafe==1.0
monotonic==1.5
python-engineio==2.3.1
python-socketio==2.0.0
six==1.11.0
Werkzeug==0.14.1
```

**Comment by miguelgrinberg** [comment]
Do you see an improvement if you increase the sleep period to, say, `0.1`?

**Comment by Jacobh2** [comment]
I tried to set a dynamic sleep on the client so that I can keep a somewhat steady speed of 240 msg/sec (pretty much minimum req for my application), so I sleep what is left of `1/240` seconds (`≈4.17ms`), which according to my logs is around `3.85` - `3.49` ms.

So with a sleep of approx x3, I'm able to loop a curl (to simlulate kublet's live-check) with only a few requests coming in at 1.3 and 2.8 seconds (meaning they would timeout).

Setting the messages per second to 10 (meaning trying to hit a sleep of `0.1`) gives me a bit more stable connection. I'm then not seeing any late responses from the curl. The only problem I see in the client is sometimes it becomes "late" due to
```
WARNING:socketIO-client: <url> [connection error] recv disconnected by SSL ([SSL: TLSV1_ALERT_PROTOCOL_VERSION] tlsv1 alert protocol version (_ssl.c:2460))
```
But I'm guessing that problem is more related to the nginx ingress that I'm using to expose the service.

What I don't understand, however, is why it seems like both the `index` method and the `on_data` methods are executed on `MainThread`? (I added some more logs). Is it possible with some setting to set the number of threads used to more than 1?

**Comment by miguelgrinberg** [comment]
@Jacobh2 you are using eventlet, which is an asynchronous framework. There is only one thread for everything, the CPU is shared among all the concurrently executing tasks cooperatively.

If a single eventlet worker can't keep up with your load requirements, then maybe you should consider running multiple workers. This is supported by Flask-SocketIO, but to make it work you need to also add a message queue such as Redis or RabbitMQ. See the docs for details.

**Comment by miguelgrinberg** [comment]
Load is not your problem, I think. The issue is that you have heavy CPU usage, which is the mortal enemy of asynchronous systems. This is all running on a single thread, so any time spent on CPU computation is time that every other task in the system is blocked. For an async system to work well you have to make sure that there are no CPU heavy tasks and the system spends most of the time doing I/O. I'm not sure if your code is actually what you show above or if that is a simplification, but in any case, with a frequency of 240 events/sec things that don't look very CPU intensive added together become a bottleneck.

You can consider scaling up to more async workers, or maybe not using async if that really doesn't fit your project well.

**Comment by Jacobh2** [comment]
Thank you for your help here, I really appreciate it!

With 'more async workers' you mean to scale up the application to more instances? Or more workers within the same instance?

I'm continuing my testing to see how I can scale up this application to handle the load!

**Comment by miguelgrinberg** [comment]
In this context I meant multiple async processes, each handling multiple concurrent requests.

**Comment by Jacobh2** [comment]
Ok, I understand!

I tested to completely change my setup and use a different websocket server (not socket io based) and a new client that also just uses websockets and was able to increase the throughput from 240 msg/s to 10k msg/s and keep it there steadily with even tighter resource limits on the cpu. The server could also handle the http requests made by kubelet for the healthchecks! 

Thank you for the help and info, we can close this question now 😃

**Comment by Jacobh2** [comment]
I'm not entirely sure how you relate async with something opposite to thread/process?

The server-side is still handling messages asynchronously, as in, acting on events when they arrive!

I'm using [this](https://github.com/dpallot/simple-websocket-server) and starting it in one thread, and then having the flask (http) part in another thread, which I start using using [this](http://flask.pocoo.org/docs/1.0/deploying/wsgi-standalone/#gevent).

I was quite surprised actually by the speed increase, and now I think that what is stopping me from increasing the number of msgs more is the network, which is exactly what I "want" to be my bottleneck

**Comment by miguelgrinberg** [comment]
An asynchronous process uses a single thread. The "threads" in this type of process are "green threads", which in the case of gevent and eventlet are implemented as greenlets. This is a good model to achieve big levels of concurrency (i.e. lots of clients). This comes at the expense of CPU processing slowing everything down.

With multithreading or multiprocessing, you can't have as many clients, and CPU processing does not have such a big effect on overall performance.

Sounds like you are using an asynchronous solution for the Flask server, but maybe not for the websocket portion?

**Comment by Jacobh2** [comment]
Ok! Thanks for clarifying!

Yes, I believe this lib I'm now using is working synchronously!

First of all, the server code that I tested with *is* the server code I provided above, meaning I don't really do anything but receiving the messages coming in, and still it couldn't handle the message rate that I require. 

Second, since I'm now able to consume messages in a **significantly** higher rate (even more than I actually need for my application), I'm able to decrease the [cpu limit](https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/#resource-requests-and-limits-of-pod-and-container) of my pod so much that I can instead increase the number of replicas to handle the number of clients that I have!
